File: server/models/Event.js
const mongoose = require('mongoose');

const EventSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  title: {
    type: String,
    required: true,
  },
  description: {
    type: String,
  },
  startTime: {
    type: Date,
    required: true,
  },
  endTime: {
    type: Date,
    required: true,
  },
  isAllDay: {
    type: Boolean,
    default: false,
  },
  type: {
    type: String,
    enum: ['task', 'meeting', 'focus', 'reminder'],
    default: 'task',
  },
  category: {
    type: String,
    enum: ['Study', 'Fitness', 'Work', 'Personal', 'Meeting', 'Other'],
    default: 'Other',
  },
  isCompleted: {
    type: Boolean,
    default: false,
  },
  referenceId: {
    type: mongoose.Schema.Types.ObjectId, // Can point to a Goal, Habit, or DailyPlan Block
  },
  referenceType: {
    type: String,
    enum: ['Goal', 'Habit', 'DailyPlan', null],
    default: null,
  },
  recurrenceRule: {
    type: String, // e.g. "FREQ=WEEKLY;BYDAY=MO,WE,FR"
  }
}, {
  timestamps: true
});

// Advanced recurring event validation and overlap detection moved to services.

module.exports = mongoose.model('Event', EventSchema);

File: flutter_client/lib/features/calendar/models/event.dart
class CalendarEvent {
  final String id;
  final String title;
  final String description;
  final DateTime startTime;
  final DateTime endTime;
  final bool isAllDay;
  final String type;
  final String category;
  final bool isCompleted;
  final String? goalId;
  final String? habitId;
  final num focusTimeSpent;
  final String? recurrenceRule;

  CalendarEvent({
    required this.id,
    required this.title,
    this.description = '',
    required this.startTime,
    required this.endTime,
    this.isAllDay = false,
    this.type = 'task',
    this.category = 'Other',
    this.isCompleted = false,
    this.goalId,
    this.habitId,
    this.focusTimeSpent = 0,
    this.recurrenceRule,
  });

  factory CalendarEvent.fromJson(Map<String, dynamic> json) {
    return CalendarEvent(
      id: json['_id'] as String,
      title: json['title'] as String,
      description: json['description'] as String? ?? '',
      startTime: DateTime.parse(json['startTime'] as String),
      endTime: DateTime.parse(json['endTime'] as String),
      isAllDay: json['isAllDay'] as bool? ?? false,
      type: json['type'] as String? ?? 'task',
      category: json['category'] as String? ?? 'Other',
      isCompleted: json['isCompleted'] as bool? ?? false,
      goalId: json['goalId'] as String?,
      habitId: json['habitId'] as String?,
      focusTimeSpent: json['focusTimeSpent'] as num? ?? 0,
      recurrenceRule: json['recurrenceRule'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'description': description,
      'startTime': startTime.toIso8601String(),
      'endTime': endTime.toIso8601String(),
      'isAllDay': isAllDay,
      'type': type,
      'category': category,
      'isCompleted': isCompleted,
      if (goalId != null) 'goalId': goalId,
      if (habitId != null) 'habitId': habitId,
      'focusTimeSpent': focusTimeSpent,
      if (recurrenceRule != null) 'recurrenceRule': recurrenceRule,
    };
  }
}
